src/lustre/dev/simulate.gleam

```diff
@@ -0,0 +1,341 @@
+/// Construct a simulated simple Lustre application. The simulation can be started
+/// with the [`start`](#start) function by providing the initial arguments for
+/// your app's `init` function.
+///
+/// DOM events and messages dispatched by effects can be simulated using the
+/// [`event`](#event) and [`messgae`](#message) functions.
+///
+pub fn simple(
+  init init: fn(args) -> model,
+  update update: fn(model, msg) -> model,
+  view view: fn(model) -> Element(msg),
+) -> App(args, model, msg) {
+  App(
+    init: fn(args) { #(init(args), effect.none()) },
+    update: fn(model, msg) { #(update(model, msg), effect.none()) },
+    view:,
+  )
+}
+
+/// Construct a simulated Lustre application. The simulation can be started
+/// with the [`start`](#start) function by providing the initial arguments for
+/// your app's `init` function.
+///
+/// DOM events and messages dispatched by effects can be simulated using the
+/// [`event`](#event) and [`messgae`](#message) functions.
+///
+/// > **Note**: simulated apps do not run any effects! You can simulate the result
+/// > of an effect by using the [`message`](#message) function, but to test side
+/// > effects you should test your application in a real environment.
+///
+pub fn application(
+  init init: fn(args) -> #(model, Effect(msg)),
+  update update: fn(model, msg) -> #(model, Effect(msg)),
+  view view: fn(model) -> Element(msg),
+) -> App(args, model, msg) {
+  App(init:, update:, view:)
+}
+
+// RUNNING SIMULATIONS ---------------------------------------------------------
+
+/// Start a simulated Lustre application. Once a simulation is running you can
+/// use the [`message`](#message) and [`event`](#event) functions to simulate
+/// events
+///
+pub fn start(app: App(args, model, msg), args: args) -> Simulation(model, msg) {
+  let #(model, _) = app.init(args)
+  let html = app.view(model)
+
+  Simulation(update: app.update, view: app.view, history: [], model:, html:)
+}
+
+/// Simulate a message sent directly to the runtime. This is often used to mimic
+/// the result of some effect you would have run in a real environment. For example,
+/// you might simulate a click event on a login button and then simulate the
+/// successful response from the server by calling this function with the message
+/// you would dispatch from the effect:
+///
+/// ```gleam
+/// import birdie
+/// import lustre/dev/simulate
+/// import lustre/dev/query
+/// import lustre/element
+///
+/// pub fn login_test() {
+///   let app = simulate.application(init:, update:, view:)
+///   let login_button = query.element(matching: query.id("login"))
+///   let user = User(name: "Lucy")
+///
+///   simulate.start(app, Nil)
+///   |> simulate.event(on: login_button, name: "click", data: [])
+///   // Simulate a successful response from the server
+///   |> simulate.message(ApiReturnedUser(Ok(user)))
+///   |> simulate.view
+///   |> element.to_readable_string
+///   |> birdie.snap("Successful login")
+/// }
+/// ```
```

Comment:
Maybe add a note saying that you likely want to use the query API to focus on the bit of the HTML you care about.